dislib/classification/csvm/base.py

```diff
@@ -355,10 +384,38 @@ def _merge_scores(*partials):
     return total_correct / total_size
 
 
-def _merge(*subsets):
-    set0 = subsets[0].copy()
+def _merge(*subsets_and_ids):
+    samples = subsets_and_ids[0].samples
+    labels = subsets_and_ids[0].labels
+    ids = subsets_and_ids[1]
 
-    for setx in subsets[1:]:
-        set0.concatenate(setx, remove_duplicates=True)
+    it = iter(subsets_and_ids[2:])
+    for subset_x, ids_x in zip(it, it):
+        set_x = (subset_x.samples, subset_x.labels, ids_x)
+        samples, labels, ids = _merge_pair((samples, labels, ids), set_x)
+    return Subset(samples, labels), ids
 
-    return set0
+
+def _merge_pair(set_0, set_1):
+    samples_0, labels_0, ids_0 = set_0
+    samples_1, labels_1, ids_1 = set_1
+    assert issparse(samples_0) == issparse(samples_1), \
+        "Cannot concatenate sparse data with non-sparse data."
+    assert (labels_0 is None) == (labels_1 is None), \
+        "Cannot concatenate labeled data with non-labeled data"
+    if issparse(samples_0):
+        samples = sp.vstack((samples_0, samples_1))
+    else:
+        samples = np.concatenate((samples_0, samples_1))
+    labels = None
+    if labels_0 is not None:
+        labels = np.concatenate((labels_0, labels_1))
+    ids = np.concatenate((ids_0, ids_1))
+    ids, uniques = np.unique(ids, return_index=True)
+    indices = np.argsort(uniques)
+    uniques = uniques[indices]
+    ids = ids[indices]
+    samples = samples[uniques]
+    if labels is not None:
+        labels = labels[uniques]
+    return samples, labels, ids
```

Comment:
To generate the new Subset (without duplicates), you can use Subset.concatenate and then slice it using subset[uniques]

Reply:
Absolutely right.

Reply:
Done.